components/feature/session/src/main/java/mozilla/components/feature/session/HistoryDelegate.kt

```diff
@@ -32,4 +36,33 @@ class HistoryDelegate(private val historyStorage: HistoryStorage) : HistoryTrack
     override suspend fun getVisited(): List<String> {
         return historyStorage.getVisited()
     }
+
+    /**
+     * Filter out unwanted URIs, such as "chrome:", "about:", etc.
+     * Ported from nsAndroidHistory::CanAddURI
+     * See https://dxr.mozilla.org/mozilla-central/source/mobile/android/components/build/nsAndroidHistory.cpp#326
+     */
+    @SuppressWarnings("ReturnCount")
+    override fun shouldStoreUri(uri: String): Boolean {
+        val parsedUri = Uri.parse(uri)
+        val scheme = parsedUri.scheme ?: return false
+
+        // Short-circuit most common schemes.
+        if (scheme == "http" || scheme == "https") {
+            return true
+        }
+
+        // Allow about about:reader uris. They are of the form:
+        // about:reader?url=http://some.interesting.page/to/read.html
+        if (uri.startsWith("about:reader")) {
+            return true
+        }
+
+        val schemasToIgnore = listOf(
+            "about", "imap", "news", "mailbox", "moz-anno", "moz-extension",
+            "view-source", "chrome", "resource", "data", "javascript", "blob"
+        )
```

Comment:
I was wondering if an "allowlist only" would be a better choice.. but seeing that this is from Fennec this seems to be fine. :)

Reply:
Not sure how well a whitelist will work here. It seems more flexible to forbid certain select schemes than to allow certain schemes.